File: src/main/java/komu/demodel/utils/ChangeListenerList.java
package komu.demodel.utils;

import java.util.concurrent.CopyOnWriteArrayList;

import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

public final class ChangeListenerList {

    private final CopyOnWriteArrayList<ChangeListener> listeners = new CopyOnWriteArrayList<ChangeListener>();
    
    public void add(ChangeListener listener) {
        Check.notNull(listener, "listener");
        
        listeners.add(listener);
    }
    
    public void stateChanged(ChangeEvent event) {
        for (ChangeListener listener : listeners)
            listener.stateChanged(event);
    }
}

File: src/main/scala/komu/demodel/utils/ChangeListenerList.scala
/*
 * Copyright (C) 2010 Juha Komulainen. All rights reserved.
 */
package komu.demodel.utils

import javax.swing.event.{ ChangeEvent, ChangeListener }

final class ChangeListenerList {

  private var listeners: List[ChangeListener] = Nil
    
  def add(listener: ChangeListener) {
    assert(listener != null)
    listeners ::= listener
  }
    
  def stateChanged(event: ChangeEvent) =
    for (listener <- listeners)
      listener.stateChanged(event)
}
